File: 2023/UICPC Regional/Round #12/solutions/stockbroker/lukas.cc
//Solution by lukasP (Lukáš Poláček)
#include <iostream>
#include <vector>
#include <algorithm>
#include <cstdio>
using namespace std;

#define rep(i,a,b) for(__typeof(b) i=a; i<(b); ++i)

typedef long long ll;
int main() {
    int n; scanf("%d", &n);
    vector<ll> a(n);
    rep(i,0,n) scanf("%lld", &a[i]);

    ll money = 100 % a[0], stocks = 100 / a[0];
    ll ma = 100;

    rep(i,1,n) {
        if (a[i] < a[i - 1]) { // Sell
            money += a[i - 1] * stocks;

            ll count = min(100000LL, money / a[i]);
            stocks = count;
            money = money - count * a[i];
        }
        ma = max(ma, money + stocks * a[i]);
    }

    cout << ma << endl;
}

File: 2023/UICPC Regional/Round #12/solutions/stockbroker/lukas.cs
//Solution by lukasP (Lukáš Poláček)
using System;
using System.IO;
using System.Linq;

class Problem
{
    struct State {
        public long money, stocks, best, previous;
        public State(long m, long s, long b, long p) {
            money = m;
            stocks = s;
            best = b;
            previous = p;
        }
    }

    static State Calculate(State state, long price) {
        long stocks = state.stocks;
        long previous = state.previous;
        if (previous > price) { // Sell
            long money = state.money + previous * stocks;
            stocks = Math.Min(100000, money / price);
            return new State(money - stocks * price, stocks, state.best, price);
        } else {
            return new State(state.money, stocks,
                Math.Max(state.best, state.money + price * stocks), price);
        }
    }

    public static void Main ()
    {
        var bs = new BufferedStream(Console.OpenStandardInput());
        StreamReader reader = new StreamReader(bs);
        int n = int.Parse(reader.ReadLine());

        long result = Enumerable.Range(0, n).Select(x => long.Parse(reader.ReadLine()))
            .Aggregate(new State(100L, 0L, 100L, 600L), Calculate).best;
        Console.WriteLine(result);
    }
}
